src/rust/engine/process_execution/src/local.rs

```diff
@@ -1,85 +1,132 @@
 extern crate tempdir;
 
-use std::io::Error;
+use boxfuture::{BoxFuture, Boxable};
+use fs::{self, PathStatGetter};
+use futures::{future, Future};
 use std::path::Path;
 use std::process::Command;
+use std::sync::Arc;
 
 use super::{ExecuteProcessRequest, ExecuteProcessResult};
 
 use bytes::Bytes;
-///
-/// Runs a command on this machine in the pwd.
-///
-pub fn run_command_locally(
-  req: ExecuteProcessRequest,
-  workdir: &Path,
-) -> Result<ExecuteProcessResult, Error> {
-  Command::new(&req.argv[0])
-    .args(&req.argv[1..])
-    .current_dir(workdir)
-    .env_clear()
-    // It would be really nice not to have to manually set PATH but this is sadly the only way
-    // to stop automatic PATH searching.
-    .env("PATH", "")
-    .envs(req.env)
-    .output()
-    .map(|output| ExecuteProcessResult {
-      stdout: Bytes::from(output.stdout),
-      stderr: Bytes::from(output.stderr),
-      exit_code: output.status.code().unwrap(),
-    })
+
+pub struct CommandRunner {
+  store: fs::Store,
+  fs_pool: Arc<fs::ResettablePool>,
+}
+
+impl CommandRunner {
+  pub fn new(store: fs::Store, fs_pool: Arc<fs::ResettablePool>) -> CommandRunner {
+    CommandRunner { store, fs_pool }
+  }
+
+  ///
+  /// Runs a command on this machine in the passed working directory.
+  ///
+  pub fn run(
+    &self,
+    req: ExecuteProcessRequest,
+    workdir: &Path,
+  ) -> BoxFuture<ExecuteProcessResult, String> {
+    let output = try_future!(
+      Command::new(&req.argv[0])
+        .args(&req.argv[1..])
+        .current_dir(workdir)
+        .env_clear()
+        // It would be really nice not to have to manually set PATH but this is sadly the only way
+        // to stop automatic PATH searching.
+        .env("PATH", "")
+        .envs(req.env)
+        .output().map_err(|e| format!("Error executing process: {:?}", e))
+    );
+
+    let output_snapshot = if req.output_files.is_empty() {
+      future::ok(fs::Snapshot::empty()).to_boxed()
+    } else {
+      // Use no ignore patterns, because we are looking for explicitly listed paths.
+      let posix_fs = Arc::new(fs::PosixFS::new(workdir, self.fs_pool.clone(), vec![]).unwrap());
```

Comment:
IMO, should handle this error. Failing here might mean that the workdir was on a custom filesystem of some kind, which is not-unlikely.

Reply:
Done.

For reference, the only reasons this can fail are:
1. The root isn't a directory (which would panic at execution time, unless it was changed between process execution and directory fetching)
2. The root could not be canonicalised (which is super weird)
3. The ignore patterns could not be parsed (but they're hard-coded as empty, so can't fail)